### PR TITLE
Boxes: Add support for Boxes

### PR DESCRIPTION
**Pull request by michaeldiamant**
Adds SDK support for Boxes introduced in https://github.com/algorand/indexer/pull/1168 and https://github.com/algorand/go-algorand/pull/4149.

### CODE REVIEW

File: .test-env
Comment (ahangsu) [inline .test-env:3]: TODO: when algorand-sdk-testing PR merged:
```suggestion
SDK_TESTING_BRANCH="master"
```

File: src/main/java/com/algorand/algosdk/v2/client/model/BoxesResponse.java
Comment (jasonpaulos) [inline src/main/java/com/algorand/algosdk/v2/client/model/BoxesResponse.java:13]: This response does not contain a `next-token` field, which is needed for indexer pagination

File: src/test/java/com/algorand/algosdk/integration/Applications.java
Comment (jasonpaulos) [inline src/test/java/com/algorand/algosdk/integration/Applications.java:313]: A less error-prone way to compare lists would be to sort and directly compare their contents

File: src/test/java/com/algorand/algosdk/integration/Applications.java
Comment (jasonpaulos) [inline src/test/java/com/algorand/algosdk/integration/Applications.java:350]: same here

File: src/main/java/com/algorand/algosdk/v2/client/algod/GetApplicationBoxByName.java
Comment (jasonpaulos) [inline src/main/java/com/algorand/algosdk/v2/client/algod/GetApplicationBoxByName.java:36]: This pattern suffers from the same drawbacks I brought up in https://github.com/algorand/js-algorand-sdk/pull/604#discussion_r975688014

Though since this code is generated, I expect we don't want to take any action here at the moment.

File: src/main/java/com/algorand/algosdk/transaction/Transaction.java
Comment (jasonpaulos) [reply src/main/java/com/algorand/algosdk/transaction/Transaction.java:1449]: Actually many app transaction fields are missing from this method, and I think it's out of scope to add support for them at the moment. We should make a separate issue for this.

File: src/main/java/com/algorand/algosdk/builder/transaction/ApplicationBaseTransactionBuilder.java
Comment (michaeldiamant) [reply src/main/java/com/algorand/algosdk/builder/transaction/ApplicationBaseTransactionBuilder.java:108]: @jasonpaulos I suspect it's a copy-paste mistake.  Thanks for catching and fixing in the child PR.

File: src/main/java/com/algorand/algosdk/builder/transaction/MethodCallTransactionBuilder.java
Comment (michaeldiamant) [reply src/main/java/com/algorand/algosdk/builder/transaction/MethodCallTransactionBuilder.java:133]: @jasonpaulos Good catch here - thanks for catching the mistake.

File: src/main/java/com/algorand/algosdk/transaction/Transaction.java
Comment (michaeldiamant) [reply src/main/java/com/algorand/algosdk/transaction/Transaction.java:798]: @jasonpaulos I opened https://github.com/algorand/java-algorand-sdk/pull/399 to hopefully improve the situation.  Can you have a look?

File: src/main/java/com/algorand/algosdk/transaction/Transaction.java
Comment (michaeldiamant) [reply src/main/java/com/algorand/algosdk/transaction/Transaction.java:1449]: @jasonpaulos I added `boxReferences` to `equals` in https://github.com/algorand/java-algorand-sdk/pull/400.  Can you have a look?

I haven't had to manage equality by-hand in some time.  Sadly, I forgot to make the update when initially doing the work.  Thanks for catching it.

Aside:  I attempted to add all members to `equals`.  The change broke `SignedTransaction` comparison in `atomic_transaction_composer` tests.  It seems unexpected.  I also assume the intent is to provide equality on all fields.  I agree with ticketing.

File: src/main/java/com/algorand/algosdk/transaction/Transaction.java
Comment (ahangsu) [reply src/main/java/com/algorand/algosdk/transaction/Transaction.java:798]: Jason's traveling today, can I take a pass on that?

File: src/main/java/com/algorand/algosdk/v2/client/model/BoxesResponse.java
Comment (michaeldiamant) [reply src/main/java/com/algorand/algosdk/v2/client/model/BoxesResponse.java:13]: @jasonpaulos It seems codegen source code is out of date.  I re-generated via https://github.com/algorand/java-algorand-sdk/pull/401.  Can you have a look?

File: src/main/java/com/algorand/algosdk/transaction/Transaction.java
Comment (jasonpaulos) [reply src/main/java/com/algorand/algosdk/transaction/Transaction.java:1445]: I created #403 to track this. Since I think we've addressed everything related to this PR, I'll resolve this thread.

**Review by jasonpaulos — APPROVED** [review]
This has my conditional approval, as long as `SDK_TESTING_BRANCH` is changed back to `master`

File: src/main/java/com/algorand/algosdk/transaction/AppBoxReference.java
Comment (ahangsu) [inline src/main/java/com/algorand/algosdk/transaction/AppBoxReference.java:33]: nit: @jasonpaulos we had a short discussion on this one before, if I remembered correctly. We introduced `ComparableBytes` to handle such problem, right? Do we want to change implementation on this one?

File: src/main/java/com/algorand/algosdk/transaction/Transaction.java
Comment (ahangsu) [inline src/main/java/com/algorand/algosdk/transaction/Transaction.java:1633]: visual reminder for hashcode discussion

File: src/main/java/com/algorand/algosdk/transaction/AppBoxReference.java
Comment (jasonpaulos) [reply src/main/java/com/algorand/algosdk/transaction/AppBoxReference.java:33]: This is a separate concern from `ComparableBytes`.

Previously the code made a `HashSet` from `AppBoxReference` instances, so `hashCode` was required to be implemented. Now that that is no longer the case, we technically don't need `hashCode`, but I think there's no harm in keeping it.

@michaeldiamant likely has more insight into this than I do.

File: src/main/java/com/algorand/algosdk/transaction/MethodCallParams.java
Comment (jasonpaulos) [reply src/main/java/com/algorand/algosdk/transaction/MethodCallParams.java:66]: Ah, I forgot to talk about this. In the most strict sense, yes this is a breaking change. However, it's strongly recommended that consumers don't use this constructor directly.

If we want to put in the effort to make this not a breaking change, I think that would ultimately be a good thing, I just question how important that is.

Comment (ahangsu) [reply src/main/java/com/algorand/algosdk/transaction/MethodCallParams.java:66]: got it got it, that is what I mean we don't consider it, though it is technically, i.e., in the most strict case, yea it is a breaking change.

Comment (michaeldiamant) [reply src/main/java/com/algorand/algosdk/transaction/MethodCallParams.java:66]: @ahangsu Unresolving - I think you're correct to point out the backwards incompatibility.  We ought to avoid the breaking change.   Thanks for considering it!

I touched base with @jasonpaulos to understand backwards compatible options and I'll open a child PR addressing the concern.  Let's leave the thread unresolved until the child PR is either closed or merged.

File: src/main/java/com/algorand/algosdk/transaction/AppBoxReference.java
Comment (michaeldiamant) [reply src/main/java/com/algorand/algosdk/transaction/AppBoxReference.java:33]: @ahangsu Thoughts on my end - Let me know if the brings enough clarity to resolve?  If not, happy to discuss live.
* I agree with @jasonpaulos's assessment that the highlighted `hashCode` override is largely orthogonal to `ComparableBytes`.
* For a class holding data (like `AppBoxReference`), I think it's a bug if `hashCode` + `equals` are _not_ overwritten.  The class _is_ intended to be used for comparison purposes.  So, I think it's correct to leave as is.
  * In Scala, the kind of class we're describing is a _case class_:  https://docs.scala-lang.org/tour/case-classes.html.    
  * As of Java 14, Java introduces the concept as _record classes_: 
 https://docs.oracle.com/en/java/javase/16/language/records.html
  * I realize the repo convention is to _omit_ `hashCode`.  Based on the above rationale, I view it as a bug:  https://github.com/algorand/java-algorand-sdk/issues/344.

Comment (ahangsu) [reply src/main/java/com/algorand/algosdk/transaction/AppBoxReference.java:33]: I see, my apologies for wronged memory. I was somewhat concerned about the implementation of the `hashcode` but now I am more clear about the situation.

I guess the most straightforward way to get around explicit definition of hashcode is to up the jvm version, at some point in the future...

File: src/main/java/com/algorand/algosdk/transaction/MethodCallParams.java
Comment (ahangsu) [reply src/main/java/com/algorand/algosdk/transaction/MethodCallParams.java:66]: got it, thanks. Waiting on a child PR now.

Comment (michaeldiamant) [reply src/main/java/com/algorand/algosdk/transaction/MethodCallParams.java:66]: Here's the child PR:  https://github.com/algorand/java-algorand-sdk/pull/408.

File: src/main/java/com/algorand/algosdk/transaction/AppBoxReference.java
Comment (michaeldiamant) [reply src/main/java/com/algorand/algosdk/transaction/AppBoxReference.java:33]: @ahangsu No need to apologize, it's a fair question.  For reference, I use the IDE (intellij) to auto-generate `equals` and `hashCode` implementations.  I find allowing the IDE to generate these methods is less error prone than a handwritten version.

File: src/main/java/com/algorand/algosdk/transaction/MethodCallParams.java
Comment (ahangsu) [reply src/main/java/com/algorand/algosdk/transaction/MethodCallParams.java]: resolving.

**Review by ahangsu — APPROVED** [review]

File: .test-env
Comment (michaeldiamant) [inline .test-env:3]:
```suggestion
SDK_TESTING_BRANCH="master"
```

**Pull request merged**
